cli/registry/client/endpoint.go

```diff
@@ -30,32 +31,36 @@ func (r repositoryEndpoint) BaseURL() string {
 }
 
 func newDefaultRepositoryEndpoint(ref reference.Named, insecure bool) (repositoryEndpoint, error) {
+	repoName := reference.TrimNamed(ref)
 	repoInfo, _ := registry.ParseRepositoryInfo(ref)
-	endpoint, err := getDefaultEndpointFromRepoInfo(repoInfo)
+	indexInfo := repoInfo.Index
+
+	endpoint, err := getDefaultEndpoint(ref, !indexInfo.Secure)
 	if err != nil {
 		return repositoryEndpoint{}, err
 	}
 	if insecure {
 		endpoint.TLSConfig.InsecureSkipVerify = true
 	}
-	return repositoryEndpoint{info: repoInfo, endpoint: endpoint}, nil
+	return repositoryEndpoint{
+		repoName:  repoName,
+		indexInfo: indexInfo,
+		endpoint:  endpoint,
+	}, nil
 }
 
-func getDefaultEndpointFromRepoInfo(repoInfo *registry.RepositoryInfo) (registry.APIEndpoint, error) {
-	var err error
-
-	options := registry.ServiceOptions{}
-	registryService, err := registry.NewService(options)
+func getDefaultEndpoint(repoName reference.Named, insecure bool) (registry.APIEndpoint, error) {
+	registryService, err := registry.NewService(registry.ServiceOptions{})
 	if err != nil {
 		return registry.APIEndpoint{}, err
 	}
-	endpoints, err := registryService.LookupPushEndpoints(reference.Domain(repoInfo.Name))
+	endpoints, err := registryService.LookupPushEndpoints(reference.Domain(repoName))
 	if err != nil {
 		return registry.APIEndpoint{}, err
 	}
 	// Default to the highest priority endpoint to return
 	endpoint := endpoints[0]
-	if !repoInfo.Index.Secure {
+	if insecure {
 		for _, ep := range endpoints {
 			if ep.URL.Scheme == "http" {
 				endpoint = ep
```

Comment:
This should probably also have a `break` here 🤔 - let's keep that for a follow-up if we think it's important.